x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -736,40 +784,74 @@ private void innerRefresh(String tokenDocId, Authentication userAuth, ActionList
                         if (invalidSource.isPresent()) {
                             onFailure.accept(invalidSource.get());
                         } else {
-                            final Map<String, Object> userTokenSource = (Map<String, Object>)
-                                ((Map<String, Object>) source.get("access_token")).get("user_token");
-                            final String authString = (String) userTokenSource.get("authentication");
-                            final Integer version = (Integer) userTokenSource.get("version");
-                            final Map<String, Object> metadata = (Map<String, Object>) userTokenSource.get("metadata");
-
-                            Version authVersion = Version.fromId(version);
-                            try (StreamInput in = StreamInput.wrap(Base64.getDecoder().decode(authString))) {
-                                in.setVersion(authVersion);
-                                Authentication authentication = new Authentication(in);
-                                UpdateRequestBuilder updateRequest =
-                                    client.prepareUpdate(SecurityIndexManager.SECURITY_INDEX_NAME, TYPE, tokenDocId)
-                                        .setDoc("refresh_token", Collections.singletonMap("refreshed", true))
-                                        .setRefreshPolicy(RefreshPolicy.WAIT_UNTIL);
-                                if (clusterService.state().nodes().getMinNodeVersion().onOrAfter(Version.V_6_7_0)) {
-                                    updateRequest.setIfSeqNo(response.getSeqNo());
-                                    updateRequest.setIfPrimaryTerm(response.getPrimaryTerm());
-                                } else {
-                                    updateRequest.setVersion(response.getVersion());
+                            if (eligibleForMultiRefresh(source)) {
+                                final Map<String, Object> refreshTokenSrc = (Map<String, Object>) source.get("refresh_token");
+                                final String supersedingTokenDocId = (String) refreshTokenSrc.get("superseded_by");
+                                logger.debug("Token document [{}] was recently refreshed, attempting to reuse [{}] for returning an " +
+                                    "access token and refresh token", tokenDocId, supersedingTokenDocId);
+                                GetRequest supersedingTokenGetRequest =
+                                    client.prepareGet(SecurityIndexManager.SECURITY_INDEX_NAME, TYPE, supersedingTokenDocId).request();
```

Comment:
The GET will issue a refresh if the document has been updated but not refreshed IIUC. Can you elaborate on what the performance impact might be and how it relates to setting `WAIT_UNTIL` on the update request?

Reply:
> As I was reading further, it was pointed to be careful with the `refresh` option as it can cause heavy load

I guess that the equivalent of `refresh=true` is if we would set `IMMEDIATE`

> So when updating a document (token) in security index, that change will not be available for searching immediately but on next refresh cycle. Since we might be getting parallel requests from Kibana it might happen that when we do Get API that will trigger refresh for the document.

This can happen I think. But we need this to happen because we want the subsequent request(s) to get the updated view of the document state

Reply:
By refreshing, we are writing lots of small lucene segments that need to be merged in the background. However, I do not feel like there is an issue with this use of Get.